ethcore/wasm/src/runtime.rs

```diff
@@ -402,449 +376,208 @@ impl<'a, 'b> Runtime<'a, 'b> {
-	}
-
-	/// Query current gas left for execution
-	pub fn gas_left(&self) -> Result<u64, UserTrap> {
-		if self.gas_counter > self.gas_limit { return Err(UserTrap::InvalidGasState); }
-		Ok(self.gas_limit - self.gas_counter)
-	}
-
-	/// Shared memory reference
-	pub fn memory(&self) -> &InterpreterMemoryInstance {
-		&*self.memory
+		self.charge(|schedule| schedule.wasm.static_address as u64)?;
+		self.memory.set(ptr, &*val)?;
+		Ok(())
 	}
 
-	fn mem_cmp(&mut self, context: InterpreterCallerContext)
-		-> Result<Option<interpreter::RuntimeValue>, InterpreterError>
-	{
-		//
-		// method signature:
-		//   fn memcmp(cx: *const u8, ct: *const u8, n: usize) -> i32;
-		//
-
-		let len = context.value_stack.pop_as::<i32>()? as u32;
-		let ct = context.value_stack.pop_as::<i32>()? as u32;
-		let cx = context.value_stack.pop_as::<i32>()? as u32;
-
-		self.charge(|schedule| schedule.wasm.mem_cmp as u64 * len as u64)?;
-
-		let ct = self.memory.get(ct, len as usize)?;
-		let cx = self.memory.get(cx, len as usize)?;
-
-		let result = unsafe {
-			memcmp(cx.as_ptr() as *const c_void, ct.as_ptr() as *const c_void, len as usize)
-		};
-
-		Ok(Some(Into::into(result)))
+	fn return_u256_ptr(&mut self, ptr: u32, val: U256) -> Result<()> {
+		let value: H256 = val.into();
+		self.charge(|schedule| schedule.wasm.static_u256 as u64)?;
+		self.memory.set(ptr, &*value)?;
+		Ok(())
 	}
 
-	fn mem_copy(&mut self, context: InterpreterCallerContext)
-		-> Result<Option<interpreter::RuntimeValue>, InterpreterError>
-	{
-		//
-		// method signature:
-		//   fn memcpy(dest: *const u8, src: *const u8, len: u32) -> *mut u8;
-		//
-
-		let len = context.value_stack.pop_as::<i32>()? as u32;
-		let src = context.value_stack.pop_as::<i32>()? as u32;
-		let dst = context.value_stack.pop_as::<i32>()? as u32;
-
-		self.charge(|schedule| schedule.wasm.mem_copy as u64 * len as u64)?;
-
-		self.memory().copy_nonoverlapping(src as usize, dst as usize, len as usize)?;
-
-		Ok(Some(Into::into(dst as i32)))
+	pub fn value(&mut self, args: RuntimeArgs) -> Result<()> {
+		let val = self.context.value;
+		self.return_u256_ptr(args.nth(0)?, val)
 	}
 
-	fn mem_move(&mut self, context: InterpreterCallerContext)
-		-> Result<Option<interpreter::RuntimeValue>, InterpreterError>
+	pub fn create(&mut self, args: RuntimeArgs) -> Result<RuntimeValue>
 	{
 		//
 		// method signature:
-		//   fn memmove(dest: *const u8, src: *const u8, len: u32) -> *mut u8;
+		//   fn create(endowment: *const u8, code_ptr: *const u8, code_len: u32, result_ptr: *mut u8) -> i32;
 		//
+		trace!(target: "wasm", "runtime: CREATE");
+		let endowment = self.u256_at(args.nth(0)?)?;
+		trace!(target: "wasm", "       val: {:?}", endowment);
+		let code_ptr: u32 = args.nth(1)?;
+		trace!(target: "wasm", "  code_ptr: {:?}", code_ptr);
+		let code_len: u32 = args.nth(2)?;
+		trace!(target: "wasm", "  code_len: {:?}", code_len);
+		let result_ptr: u32 = args.nth(3)?;
+		trace!(target: "wasm", "result_ptr: {:?}", result_ptr);
```

Comment:
I think this code will more legible if we won't interleave argument decoding code with displaying it, as we do in `do_call`.
At least, I'm wondering, should we synchronize them?

Reply:
why?